### PR TITLE
Clarify blog post about extension methods

### PR DESCRIPTION
**Pull request by LPTK**

### CODE REVIEW

File: docs/blog/_posts/2019-01-21-12th-dotty-milestone-release.md [hunk 1]
```diff
@@ -35,19 +35,41 @@ This is our 12th scheduled release according to our
...
 case class Circle(x: Double, y: Double, radius: Double)
 
 def (c: Circle) circumference: Double = c.radius * math.Pi * 2
 ```
 
-Read the [relevant documentation](https://dotty.epfl.ch/docs/reference/other-new-features/extension-methods.html) about generic extension methods, higher-kinded extension methods and more.
+Extension methods were previously encoded in a rather roundabout way via the implicit class pattern.
+Such encoding required a lot of boilerplate, especially when defining type classes.
+In Dotty, this is no longer the case,
+and type classes with infix syntax become very straightforward to define!
```
Comment (biboudis) [inline docs/blog/_posts/2019-01-21-12th-dotty-milestone-release.md]: 👍

File: docs/blog/_posts/2019-01-21-12th-dotty-milestone-release.md [hunk 2]
```diff
@@ -35,19 +35,41 @@ This is our 12th scheduled release according to our
...
+For example, consider:
```
Comment (biboudis) [inline docs/blog/_posts/2019-01-21-12th-dotty-milestone-release.md]: I think we dive in the topic of extension methods & type classes directly by skipping generic extension methods. Yes, it has been less straightforward before, but extension methods are a general language feature (others coming from other ecosystems do not use them for type classes only). This is why I left more elaborate examples out.

Comment (LPTK) [reply docs/blog/_posts/2019-01-21-12th-dotty-milestone-release.md]: By _elaborate_, you mean that it has a type parameter? I think most Scala users (even beginners) won't have a problem understanding this code snippet. But of course I can change `ListSemigroup` to `StringSemigroup` for example, if that makes it simpler.

File: docs/blog/_posts/2019-01-21-12th-dotty-milestone-release.md [hunk 3]
```diff
@@ -35,19 +35,41 @@ This is our 12th scheduled release according to our
...
 
 We are excited to announce that extension methods are now offered through dedicated language support!
-Extension methods allow one to add methods to a type after the type is defined and up until now they were encoded through the powerful mechanism of implicits. 
-Previously, one had to place the desired method in the trait with the receiver object to extend, as part of the parameter list. 
-The infix-ness of the method was being restored by manually writing the implicit class, resolving `implicitly` that extended object. 
-
-Now, in Dotty we can express an extension method by writing the method and prepending its name with the type to be extended. 
+Extension methods allow one to add methods to a type after the type is defined.
+This is done by writing a method with a parameter for the type to be extended
+_on the right-hand side_ of the method name:
```
Comment (julienrf) [inline docs/blog/_posts/2019-01-21-12th-dotty-milestone-release.md]: You mean on the left-hand side?

**Comment by LPTK** [comment]
IMHO the significance of Dotty's extension methods is _not_ that they use less boilerplate than implicit classes. Indeed, the reason the community has been so excited about them is that they allow straightforward type class definitions with infix syntax, which is a huge improvement. People reading this need to know that we didn't just "copy C#", and that extension methods in Dotty are way more general and useful (almost a game changer, I'd say).

File: docs/blog/_posts/2019-01-21-12th-dotty-milestone-release.md (hunk 3 above)
Comment (LPTK) [reply docs/blog/_posts/2019-01-21-12th-dotty-milestone-release.md]: Oops, good catch!

**Comment by biboudis** [comment]
My observation is just that we shouldn't jump from the simplest case of `Circle` into type classes' definitions with natural syntax without passing first from extension methods defined in a trait and generic extension methods. Essentially we would need to reproduce the documentation Community members know, new users are motivated when they map new concepts to stuff they already know ;-)

However since I am a strong advocate of examples I wouldn't contest to remove it. 

WDYT about putting a minor list below the `Circle` example saying something like?

>Extension methods can be:
>- visible under a simple name (like in the `Circle` example above)
>- enclosed in a trait and enabled by bringing them in scope
>- polymorphic 
>- improve type class definitions with natural/infix syntax

**Comment by LPTK** [comment]
> I wouldn't contest to remove it.

I'm not sure to understand if you're proposing to remove the type class example or not.

> WDYT about putting a minor comment below the `Circle` example

Sure, why not. But I'm not sure why list "polymorphic". All methods can be polymorphic, it's not a special thing of extension methods. Or maybe you mean something else? Also, why make a difference between "visible under a simple name" and "enabled by bringing them in scope"? And why "enclosed in a trait" when they can be enclosed in anything (trait, class, object...)? Wouldn't it be simpler to just say "enabled when in scope"?

**Comment by biboudis** [comment]
> I'm not sure to understand if you're proposing to remove the type class example or not.
 
Not anymore. Don't remove it. Just introduce it better with a small comment like the one above and we are good.

> Wouldn't it be simpler to just say "enabled when in scope"?

Sound's good, it is also different with other practices after all ;-)

**Comment by LPTK** [comment]
@biboudis done!

Please, let's try to merge this sooner than later, while the blog post is still being read by people; because as it is now, it contains a pretty obscure and incomplete presentation of extension methods.

**Comment by biboudis** [comment]
Looks good. Merged. 

Obscurity wasn't intended (you rephrased sentences perfectly!), incompleteness was. 
It is the announcement of a feature. The rest are discussed at the documentation.

**Pull request merged**